java-squid/src/main/java/org/sonar/java/collections/AVLTree.java

```diff
@@ -0,0 +1,388 @@
+import java.util.Objects;
+
+/**
+ * AVL Tree.
+ *
+ * https://en.wikipedia.org/wiki/AVL_tree
+ *
+ * @author Evgeny Mandrikov
+ */
+public abstract class AVLTree<K, V> implements PMap<K, V>, PSet<K> {
+
+  /**
+   * @return empty tree
+   */
+  @SuppressWarnings("unchecked")
+  public static <K, V> AVLTree<K, V> create() {
+    return EMPTY;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public AVLTree<K, V> add(K e) {
+    return put(e, e, this);
+  }
+
+  @Override
+  public boolean contains(K k) {
+    return get(k) != null;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public AVLTree<K, V> put(K key, V value) {
+    Preconditions.checkNotNull(key);
+    Preconditions.checkNotNull(value);
+    return put(key, value, this);
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public AVLTree<K, V> remove(K key) {
+    Preconditions.checkNotNull(key);
+    return remove(key, this);
+  }
+
+  @SuppressWarnings("unchecked")
+  @Nullable
+  @Override
+  public V get(K key) {
+    Preconditions.checkNotNull(key);
+    AVLTree t = this;
+    while (!t.isEmpty()) {
+      int c = KEY_COMPARATOR.compare(key, t.key());
+      if (c == 0) {
+        return (V) t.value();
+      } else if (c < 0) {
+        t = t.left();
+      } else {
+        t = t.right();
+      }
+    }
+    return null;
+  }
+
+  @Override
+  public void forEach(PSet.Consumer<K> action) {
+    forEach(this, action);
+  }
+
+  @SuppressWarnings("unchecked")
+  private void forEach(AVLTree t, PSet.Consumer<K> action) {
+    if (t.isEmpty()) {
+      return;
+    }
+    action.accept((K) t.key());
+    forEach(t.left(), action);
+    forEach(t.right(), action);
+  }
+
+  @Override
+  public void forEach(PMap.Consumer<K, V> consumer) {
+    forEach(this, consumer);
+  }
+
+  @SuppressWarnings("unchecked")
+  private void forEach(AVLTree t, PMap.Consumer<K, V> action) {
+    if (t.isEmpty()) {
+      return;
+    }
+    action.accept((K) t.key(), (V) t.value());
+    forEach(t.left(), action);
+    forEach(t.right(), action);
+  }
+
+  public interface Visitor<K, V> {
+    void visit(K key, V value);
+  }
+
+  protected abstract AVLTree left();
+
+  protected abstract AVLTree right();
+
+  protected abstract Object key();
+
+  protected abstract Object value();
+
+  protected abstract int height();
+
+  private static AVLTree put(Object key, Object value, AVLTree t) {
+    if (t.isEmpty()) {
+      return createNode(t, key, value, t);
+    }
+    int result = KEY_COMPARATOR.compare(key, t.key());
+    if (result == 0) {
+      if (value.equals(t.value())) {
+        return t;
+      }
+      return createNode(t.left(), key, value, t.right());
+    } else if (result < 0) {
+      AVLTree left = put(key, value, t.left());
+      if (left == t.left()) {
+        return t;
+      }
+      return balance(left, t.key(), t.value(), t.right());
+    } else {
+      AVLTree right = put(key, value, t.right());
+      if (right == t.right()) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)